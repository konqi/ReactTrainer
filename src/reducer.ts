enum ActionType {
  ADD_TRAINEE
}

interface Action {
  type: ActionType;
  payload: Trainee;
}

interface Trainee {
  name: string;
  price: number;
}

interface ApplicationState {
  trainees: Trainee[];
}

function reducer(state: ApplicationState, action: Action): ApplicationState {
  switch (action.type) {
    case ActionType.ADD_TRAINEE:
      const newTrainees = [...state.trainees, action.payload];
      return { ...state, trainees: newTrainees };
    default:
      throw new Error();
  }
}
