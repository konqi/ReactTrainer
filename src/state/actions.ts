export enum ActionType {
  ADD_TRAINEE
}
