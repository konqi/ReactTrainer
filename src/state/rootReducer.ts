import {ApplicationState} from '.'
import {FSA} from '../types/FSA'
import {reduceTrainees} from './trainees'

export const rootReducer = (
  state: ApplicationState,
  action: FSA<any, any>
): ApplicationState => ({
  trainees: reduceTrainees(state.trainees, action),
})
