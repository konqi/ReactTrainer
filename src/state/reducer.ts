import { ActionType } from "./actions";

export interface Action {
  type: ActionType;
  payload: Trainee;
}

export interface Trainee {
  name: string;
  price: number;
}

export interface ApplicationState {
  trainees: Trainee[];
}

export const initialApplicationState: ApplicationState = {
  trainees: []
};

export const reducer = (
  state: ApplicationState,
  action: Action
): ApplicationState => {
  switch (action.type) {
    case ActionType.ADD_TRAINEE:
      const newTrainees = [...state.trainees, action.payload];
      return { ...state, trainees: newTrainees };
    default:
      throw new Error();
  }
};
