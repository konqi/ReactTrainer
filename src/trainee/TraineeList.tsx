import {
  Fab,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from '@material-ui/core'
import {createStyles, makeStyles, Theme} from '@material-ui/core/styles'
import {Add as AddIcon} from '@material-ui/icons'
import * as React from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {ApplicationState} from '../state'
import {createShowTraineeDetailsIntend} from '../state/intends/UserIntend'
import {createUiNavigateAction} from '../state/ui/uiActions'
import {Page} from '../types/page'
import {Trainee} from '../types/Trainee'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      margin: `${theme.spacing(3)}px 0`,
      position: 'relative',
      display: 'flex',
      flexDirection: 'column',
      flexWrap: 'nowrap',
    },
    table: {
      minWidth: '85%',
    },
    fabButton: {
      margin: `${theme.spacing(3)}px auto`,
    },
  })
)

export const TraineeList: React.FC = () => {
  const classes = useStyles()
  const trainees = useSelector<ApplicationState, Trainee[]>(state =>
    Object.values(state.trainees)
  )
  const dispatch = useDispatch()
  const openTrainee = (traineeId: string) => {
    dispatch(createShowTraineeDetailsIntend(traineeId))
  }
  const newTrainee = () => {
    dispatch(createUiNavigateAction(Page.Create))
  }

  return (
    <React.Fragment>
      <Paper className={classes.root}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell>Name</TableCell>
              <TableCell align="right">Preis</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {trainees.map(trainee => (
              <TableRow
                key={trainee.id}
                hover
                onClick={() => openTrainee(trainee.id)}
              >
                <TableCell component="th" scope="row">
                  {trainee.name}
                </TableCell>
                <TableCell align="right">{trainee.price}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
        <Fab
          color="secondary"
          aria-label="Add"
          className={classes.fabButton}
          onClick={newTrainee}
        >
          <AddIcon />
        </Fab>
      </Paper>
    </React.Fragment>
  )
}
