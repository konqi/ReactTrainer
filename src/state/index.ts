import {TraineeState, TraineeFSAs} from './trainees/types'
import {initialTraineeState} from './trainees'
import {UiState, UiFSAs} from './ui/types'
import {TrainingFSAs} from './trainings/types'
import {initialUiState} from './ui/uiReducer'
import {createStore, applyMiddleware} from 'redux'
import {rootReducer} from './rootReducer'
import createSagaMiddleware from 'redux-saga'
import {rootSaga} from './sagas/rootSaga'

export interface ApplicationState {
  trainees: TraineeState
  ui: UiState
}

export const initialApplicationState: ApplicationState = {
  trainees: initialTraineeState,
  ui: initialUiState,
}

export type Action = TraineeFSAs | TrainingFSAs | UiFSAs

const sagaMiddleware = createSagaMiddleware()

export const store = createStore(
  rootReducer,
  initialApplicationState,
  applyMiddleware(sagaMiddleware)
)

sagaMiddleware.run(rootSaga)
