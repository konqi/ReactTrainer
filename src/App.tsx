import {
  AppBar,
  BottomNavigation,
  BottomNavigationAction,
  Container,
  createStyles,
  CssBaseline,
  makeStyles,
  Theme,
  Toolbar,
  Typography,
} from '@material-ui/core'
import {
  CalendarToday as CalendarIcon,
  Money as MoneyIcon,
  People as PeopleIcon,
  PersonAdd as PersonAddIcon,
} from '@material-ui/icons'
import React from 'react'
import {Provider, useSelector, useDispatch} from 'react-redux'
import './App.css'
import {ApplicationState, store} from './state'
import {UiState} from './state/ui/types'
import {CreateTrainee, TraineeDetails, TraineeList} from './trainee'
import {Page} from './types/page'
import {createUiNavigateAction} from './state/ui/uiActions'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      flexDirection: 'column',
      flexWrap: 'nowrap',
      justifyContent: 'space-between',
      width: '100vw',
      height: '100vh',
    },
    body: {
      flexGrow: 1,
    },
    title: {},
  })
)

const App: React.FC = () => {
  const classes = useStyles()
  const ui = useSelector<ApplicationState, UiState>(state => state.ui)
  const dispatch = useDispatch()
  const handleNavigation = (event: any, value: Page) => {
    dispatch(createUiNavigateAction(value))
  }

  return (
    <React.Fragment>
      <CssBaseline />
      <div className={classes.root}>
        <AppBar position="static">
          <Toolbar>
            <Typography variant="h6" className={classes.title}>
              ReactTrainer
            </Typography>
          </Toolbar>
        </AppBar>
        <Container maxWidth="sm" className={classes.body}>
          {ui.page === Page.Trainees && <TraineeList />}
          {ui.page === Page.Create && <CreateTrainee />}
          {ui.page === Page.Trainee && (
            <TraineeDetails {...ui.params as {traineeId: string}} />
          )}
        </Container>
        <Container maxWidth="sm">
          <BottomNavigation showLabels onChange={handleNavigation}>
            <BottomNavigationAction
              label="Personen"
              icon={<PeopleIcon />}
              value={Page.Trainees}
            />
            <BottomNavigationAction
              label="Person anlegen"
              icon={<PersonAddIcon />}
              value={Page.Create}
            />
            <BottomNavigationAction
              label="Finanzen"
              icon={<MoneyIcon />}
              value={Page.Finance}
            />
            <BottomNavigationAction
              label="Kalender"
              icon={<CalendarIcon />}
              value={Page.Calendar}
            />
          </BottomNavigation>
        </Container>
      </div>
    </React.Fragment>
  )
}

const StateWrapper: React.FC = () => (
  <Provider store={store}>
    <App />
  </Provider>
)

export default StateWrapper
