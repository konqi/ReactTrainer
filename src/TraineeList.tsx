import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow
} from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import * as React from "react";
import { useContext } from "react";
import { ApplicationContext } from "./App";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: "100%",
      marginTop: theme.spacing(3),
      overflowX: "auto"
    },
    table: {
      minWidth: "85%"
    }
  })
);

export const TraineeList = () => {
  const classes = useStyles();

  const { state } = useContext(ApplicationContext);

  return (
    <Paper className={classes.root}>
      <Table className={classes.table}>
        <TableHead>
          <TableRow>
            <TableCell>Name</TableCell>
            <TableCell align="right">Preis</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {state!.trainees.map(trainee => (
            <TableRow key={trainee.name}>
              <TableCell component="th" scope="row">
                {trainee.name}
              </TableCell>
              <TableCell align="right">{trainee.price}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </Paper>
  );
};
