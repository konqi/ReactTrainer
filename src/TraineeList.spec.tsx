import * as React from "react";
import { TraineeList } from "./TraineeList";
import { default as TestRenderer } from "react-test-renderer";
import { ApplicationContext } from "./App";
import { ApplicationState } from "./state/reducer";

describe("TraineeList", () => {
  test.each`
    description      | trainees
    ${"no trainee"}  | ${[]}
    ${"one trainee"} | ${[{ name: "Tester", price: 1500 }]}
  `(
    "should show TraineeList with $description",
    ({ description, trainees }) => {
      const state: ApplicationState = {
        trainees: trainees
      };

      const something = TestRenderer.create(
        <ApplicationContext.Provider value={{ state }}>
          <TraineeList />
        </ApplicationContext.Provider>
      );

      expect(something).toMatchSnapshot();
    }
  );
});
