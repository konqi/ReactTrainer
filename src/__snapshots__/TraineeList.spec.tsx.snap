// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraineeList should show TraineeList with no trainee 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 makeStyles-root-1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root makeStyles-table-2"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
          scope="col"
        >
          Preis
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    />
  </table>
</div>
`;

exports[`TraineeList should show TraineeList with one trainee 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 makeStyles-root-1 MuiPaper-rounded"
>
  <table
    className="MuiTable-root makeStyles-table-2"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Name
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
          scope="col"
        >
          Preis
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
    >
      <tr
        className="MuiTableRow-root"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          scope="row"
        >
          Tester
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
        >
          1500
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
